### PR TITLE
[WIP][SPARK-23435][SPARKR][TESTS] Update testthat to >= 2.0.0

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

- Update `testthat` to >= 2.0.0
- Replace of `testthat:::run_tests` with `testthat:::test_package_dir`
- Add trivial assertions for tests, without any expectations, to avoid skipping.
- Update related docs.

### Why are the changes needed?

`testthat` version has been frozen by [SPARK-22817](https://issues.apache.org/jira/browse/SPARK-22817) / https://github.com/apache/spark/pull/20003, but 1.0.2 is pretty old, and we shouldn't keep things in this state forever.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Existing tests.

**Note** 

This change leads to 

> Error: Test failed: 'cleanClosure on R functions'
> * node stack overflow

in `test_utils` `cleanClosure` test, which requires further investigation. 

CC @HyukjinKwon

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @felixcheung

**Comment by HyukjinKwon** [comment]
If it works, we should install the latest `testthat` in Jenkins machine. cc @shaneknapp

**Review by HyukjinKwon — COMMENTED** [review]
@zero323, thanks for putting your efforts here. As long as the tests and CRAN check pass, it should be good to go.

**Comment by zero323** [comment]
Hmmm... So it looks like it actually yields some additional errors on Windows.  

```
test_sparkSQL.R:298: skip: create DataFrame from RDD
Reason: Hive is not build with SparkSQL, skipped
test_sparkSQL.R:862: failure: collect() support Unicode characters
rdf$name[1] not equal to markUtf8("<U+C548><U+B155><U+D558><U+C138><U+C694>").
1/1 mismatches
x[1]: "<U+C548><U+B155><U+D558><U+C138><U+C694>"
y[1]: "<U+C548><U+B155><U+D558><U+C138><U+C694>"
test_sparkSQL.R:863: failure: collect() support Unicode characters
rdf$name[2] not equal to markUtf8("<U+60A8><U+597D>").
1/1 mismatches
x[1]: "<U+60A8><U+597D>"
y[1]: "<U+60A8><U+597D>"
test_sparkSQL.R:864: failure: collect() support Unicode characters
rdf$name[3] not equal to markUtf8("<U+3053><U+3093><U+306B><U+3061><U+306F>").
1/1 mismatches
x[1]: "<U+3053><U+3093><U+306B><U+3061><U+306F>"
y[1]: "<U+3053><U+3093><U+306B><U+3061><U+306F>"
test_sparkSQL.R:865: failure: collect() support Unicode characters
rdf$name[4] not equal to markUtf8("Xin ch�o").
1/1 mismatches
x[1]: NA
y[1]: "Xin ch�o"
test_sparkSQL.R:868: failure: collect() support Unicode characters
collect(where(df1, df1$name == markUtf8("<U+60A8><U+597D>")))$name not equal to markUtf8("<U+60A8><U+597D>").
Modes: list, character
Length mismatch: comparison on first 0 components
test_sparkSQL.R:1325: skip: test HiveContext
Reason: Hive is not build with SparkSQL, skipped
```

Especially the last two are worrying. I'll investigate once I put my hands on some Windows environment.

**Comment by zero323** [comment]
I've checked some of the failures, and it is pretty clear that these tests must have been silenced somehow.

- `node stack overflow` in `cleanClosure` is an obvious bug and it is trivial to reproduce outside tests - I've opened [SPARK-30629](https://issues.apache.org/jira/browse/SPARK-30629) to track this one.
- Seems like mismatches in `collect() support Unicode characters` are caused by Windows / R encoding quirks. As is right now `lines` are not even valid as a JSON input

    ```r
    library(SparkR)
    SparkR::sparkR.session()
    Sys.info()
    #           sysname           release           version 
    #         "Windows"      "Server x64"     "build 17763" 
    #          nodename           machine             login 
    # "WIN-5BLT6Q610KH"          "x86-64"   "Administrator" 
    #              user    effective_user 
    #   "Administrator"   "Administrator" 
    
    Sys.getlocale()
    
    # [1] "LC_COLLATE=English_United States.1252;LC_CTYPE=English_United States.1252;LC_MONETARY=English_United States.1252;LC_NUMERIC=C;LC_TIME=English_United States.1252"
    
    lines <- c("{\"name\":\"안녕하세요\"}",
               "{\"name\":\"您好\", \"age\":30}",
               "{\"name\":\"こんにちは\", \"age\":19}",
               "{\"name\":\"Xin chào\"}")
    
    jsonPath <- tempfile(pattern = "sparkr-test", fileext = ".tmp")
    writeLines(lines, jsonPath)
    
    df <- read.df(jsonPath, "json")
    
    
    printSchema(df)
    # root
    #  |-- _corrupt_record: string (nullable = true)
    #  |-- age: long (nullable = true)
    #  |-- name: string (nullable = true)
    
    head(df)
    #              _corrupt_record age                                     name
    # 1                       <NA>  NA <U+C548><U+B155><U+D558><U+C138><U+C694>
    # 2                       <NA>  30                         <U+60A8><U+597D>
    # 3                       <NA>  19 <U+3053><U+3093><U+306B><U+3061><U+306F>
    # 4 {"name":"Xin ch<U+FFFD>o"}  NA                                     <NA>
    ```


Let's see what else surfaces downstream.

**Comment by shaneknapp** [comment]
NOOOOOOOOOOOO!!!!!!!!11111one

i really don't want to touch the extremely fragile R "ecosystem" unless i have to.  but, this looks like i probably should...  :(

i'll experiment w/upgrading `testthat` to 2.0.0 on my staging server and cross my fingers that it doesn't open dependency hell.

we also might need to backport this change for the 2.4 branch if it causes those builds to fail.

**Comment by shaneknapp** [comment]
welp, that went smoothly...  which makes me nervous lol.  :)

```
* installing *source* package ‘testthat’ ...
** package ‘testthat’ successfully unpacked and MD5 sums checked
** libs
gcc -std=gnu99 -I/usr/share/R/include -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o
gcc -std=gnu99 -I/usr/share/R/include -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c reassign.c -o reassign.o
g++  -I/usr/share/R/include -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c test-catch.cpp -o test-catch.o
g++  -I/usr/share/R/include -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c test-example.cpp -o test-example.o
g++  -I/usr/share/R/include -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c test-runner.cpp -o test-runner.o
g++ -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o testthat.so init.o reassign.o test-catch.o test-example.o test-runner.o -L/usr/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/testthat/libs
** R
** inst
** tests
** preparing package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded
* DONE (testthat)
```

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 1]
```diff
@@ -848,24 +848,31 @@ test_that("collect() and take() on a DataFrame return the same number of rows an
 })
 
 test_that("collect() support Unicode characters", {
-  lines <- c("{\"name\":\"안녕하세요\"}",
-             "{\"name\":\"您好\", \"age\":30}",
-             "{\"name\":\"こんにちは\", \"age\":19}",
-             "{\"name\":\"Xin chào\"}")
+  jsonPath <- file.path(
```
Comment (zero323) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: The alternative seems to be to be explicit. Not sure if that's preferred here or not.

```
test_that("collect() support Unicode characters", {

  lines <- markUtf8(c(
    '{"name": "안녕하세요"}',
    '{"name": "您好", "age": 30}',
    '{"name": "こんにちは", "age": 19}',
    '{"name": "Xin ch\xc3\xa0o"}'
  ))

  jsonPath <- tempfile(pattern = "sparkr-test", fileext = ".tmp")
  writeLines(lines, jsonPath, useBytes = TRUE)

  expected <- regmatches(lines, regexec('(?<="name": ").*?(?=")', lines, perl = TRUE))

  df <- read.df(jsonPath, "json")
  rdf <- collect(df)
  expect_true(is.data.frame(rdf))

  rdf$name <- markUtf8(rdf$name)
  expect_equal(rdf$name[1], expected[[1]])
  expect_equal(rdf$name[2], expected[[2]])
  expect_equal(rdf$name[3], expected[[3]])
  expect_equal(rdf$name[4], expected[[4]])

  df1 <- createDataFrame(rdf)
  expect_equal(
    collect(
      where(df1, df1$name == expected[[2]])
    )$name,
    expected[[2]]
  )
})
```

File: R/pkg/tests/fulltests/data/test_utils_utf.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{"name": "안녕하세요"}
+{"name": "您好", "age": 30}
+{"name": "こんにちは", "age": 19}
+{"name": "Xin chào"}
```
Comment (zero323) [inline R/pkg/tests/fulltests/data/test_utils_utf.json]: I cannot really figure why, but `\\N{LATIN SMALL LETTER A WITH GRAVE}` seems to mess up things on Windows.

File: R/pkg/tests/fulltests/test_context.R [hunk 3]
```diff
@@ -84,6 +84,7 @@ test_that("rdd GC across sparkR.stop", {
   countRDD(rdd3)
   countRDD(rdd4)
   sparkR.session.stop()
+  expect_true(TRUE)
```
Comment (zero323) [inline R/pkg/tests/fulltests/test_context.R]: Testhat 2.0 skips tests which have no expectations.

File: R/pkg/tests/fulltests/test_utils.R [hunk 4]
```diff
@@ -89,7 +89,10 @@ test_that("cleanClosure on R functions", {
     lapply(x, g) + 1  # Test for capturing function call "g"'s closure as a argument of lapply.
     l$field[1, 1] <- 3  # Test for access operators `$`.
     res <- defUse + l$field[1, ]  # Test for def-use chain of "defUse", and "" symbol.
-    f(res)  # Test for recursive calls.
+    # Enable once SPARK-30629 is fixed
```
Comment (zero323) [inline R/pkg/tests/fulltests/test_utils.R]: That's clear bug, trivial to reproduce ([SPARK-30629](https://issues.apache.org/jira/browse/SPARK-30629)). Not sure why I didn't fail on `testthat` 1.x.

File: R/pkg/tests/run-all.R [hunk 5]
```diff
@@ -15,6 +15,11 @@
 # limitations under the License.
 #
 
+# TODO Remove before merging SPARK-23435
+if (grepl("^1\\.\\d+\\.\\d+$", installed.packages()["testthat", "Version"])) {
```
Comment (zero323) [inline R/pkg/tests/run-all.R]: Just not to fail all tests on Jenkins. To be removed.

Comment (zero323) [reply R/pkg/tests/run-all.R]: Doesn't seem to work anyway...

```
Warning message:
package ���testthat��� is not available (for R version 3.1.1) 
```

**Comment by zero323** [comment]
@shaneknapp

> NOOOOOOOOOOOO!!!!!!!!11111one

> i really don't want to touch the extremely fragile R "ecosystem" unless i have to. but, this looks like i probably should... :(

If you don't mind me asking ‒ have you thought about moving R environments to ananconda. If I recall correctly some discussions on dev, you we're pretty enthusiastic about using it with Python.

**Comment by HyukjinKwon** [comment]
@shaneknapp, I know what's like (https://github.com/apache/spark/pull/27328#issuecomment-578262785) .. lol .. this has been the pain in the ass in the dev env because nowadays many R libraries have a higher version of `testthat` as a dependency

**Comment by shaneknapp** [comment]
> @shaneknapp
> 
> > NOOOOOOOOOOOO!!!!!!!!11111one
> 
> > i really don't want to touch the extremely fragile R "ecosystem" unless i have to. but, this looks like i probably should... :(
> 
> If you don't mind me asking ‒ have you thought about moving R environments to ananconda. If I recall correctly some discussions on dev, you we're pretty enthusiastic about using it with Python.

i just tried moving to anaconda and it doesn't seem to support versions high enough of many of the packages that we require (for example, testthat is 1.0.7 or something IIRC).

**Comment by zero323** [comment]
@shaneknapp 

> i just tried moving to anaconda and it doesn't seem to support versions high enough of many of the packages that we require (for example, testthat is 1.0.7 or something IIRC).

2.3.1 [is available](https://github.com/conda-forge/r-testthat-feedstock) from forge, and if nothing changed, the remaining ones are as well. I should have working environment files stashed somewhere, if that helps, but there probably in some place I won't be able to access for a week or two.

**Comment by shaneknapp** [comment]
a week of two is fine, and i'm more than happy to play w/an environment
file and see if that works.

**Pull request closed without merging**